src/main/java/com/linecorp/armeria/common/ServiceInvocationContextAwarePromise.java

```diff
@@ -0,0 +1,159 @@
+package com.linecorp.armeria.common;
+
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Stream;
+
+import io.netty.util.concurrent.Future;
+import io.netty.util.concurrent.GenericFutureListener;
+import io.netty.util.concurrent.Promise;
+
+final class ServiceInvocationContextAwarePromise<T> implements Promise<T> {
+
+    private final ServiceInvocationContext context;
+    private final Promise<T> delegate;
+
+    ServiceInvocationContextAwarePromise(ServiceInvocationContext context, Promise<T> delegate) {
+        this.context = context;
+        this.delegate = delegate;
+    }
+
+    @Override
+    public Promise<T> setSuccess(T result) {
+        return delegate.setSuccess(result);
+    }
+
+    @Override
+    public boolean trySuccess(T result) {
+        return delegate.trySuccess(result);
+    }
+
+    @Override
+    public Promise<T> setFailure(Throwable cause) {
+        return delegate.setFailure(cause);
+    }
+
+    @Override
+    public boolean tryFailure(Throwable cause) {
+        return delegate.tryFailure(cause);
+    }
+
+    @Override
+    public boolean setUncancellable() {
+        return delegate.setUncancellable();
+    }
+
+    @Override
+    public Promise<T> addListener(
+            GenericFutureListener<? extends Future<? super T>> listener) {
+        return delegate.addListener(context.makeContextAware(listener));
+    }
+
+    @SuppressWarnings({"unchecked", "rawtypes"})
+    @Override
+    public Promise<T> addListeners(
+            GenericFutureListener<? extends Future<? super T>>... listeners) {
+        return delegate.addListeners(
+                Stream.of(listeners)
+                      .map(context::makeContextAware)
+                      .toArray(GenericFutureListener[]::new));
+    }
+
+    @Override
+    public Promise<T> removeListener(
+            GenericFutureListener<? extends Future<? super T>> listener) {
+        return delegate.removeListener(listener);
```

Comment:
This will not work because we added a wrapped listener and a user will specify a unwrapped listener here. The default Promise implementation performs identity comparison, so overriding equals/hashCode() won't work as expected.